hal/i2c_api.h

```diff
@@ -61,91 +76,170 @@ typedef struct i2c_s i2c_t;
  *
- *  @param obj The I2C object
+ *  @param obj        The I2C object
+ *  @param frequency  Frequency in Hz
+ *
+ *  @returns The actual frequency that the peripheral will be generating to
+ *           allow a user adjust its strategy in case the target cannot be
+ *           reached.
  */
-int  i2c_start(i2c_t *obj);
+uint32_t i2c_frequency(i2c_t *obj, uint32_t frequency);
 
-/** Send STOP command
+/** Enable or disable clock stretching for the I2C peripheral.
  *
- *  @param obj The I2C object
+ *  The behaviour is undefined unless `obj` points to a valid 'i2c_t' object
+ *  and the target supports configuring clock stretching, indicated by the
+ *  'supports_clock_stretching' attribute returned by the 'i2c_get_capabilities'
+ *  function.
+ *
+ * @param obj     The I2C object
+ * @param enabled If 'true' enable clock stretching on the given I2C peripheral,
+ *                otherwise disable it.
  */
-int  i2c_stop(i2c_t *obj);
+void i2c_set_clock_stretching(i2c_t *obj, const bool enabled);
 
-/** Blocking reading data
+/** Send START command
  *
- *  @param obj     The I2C object
- *  @param address 7-bit address (last bit is 1)
- *  @param data    The buffer for receiving
- *  @param length  Number of bytes to read
- *  @param stop    Stop to be generated after the transfer is done
- *  @return Number of read bytes
+ *  @param obj The I2C object.
  */
-int i2c_read(i2c_t *obj, int address, char *data, int length, int stop);
+void i2c_start(i2c_t *obj);
+
+/** Send STOP command
+ *
+ *  @param obj The I2C object
+ */
+void i2c_stop(i2c_t *obj);
 
 /** Blocking sending data
+ *
+ *  This function transmits data, when the peripheral is configured as Master to
+ *  the selected slave, and when configured as Slave transmits data to the
+ *  Master.
+ *
+ *  This function is blocking, it will return when the transfer is complete or a
+ *  timeout event is triggered. The number of bytes transmitted is returned by
+ *  the function after the operation is completed. Transmit operation cannot be
+ *  cancelled or aborted.
+ *
+ *  The data buffer must stay allocated during the duration of the transfer and
+ *  the contents must not be modified. The value of the specified `address` is
+ *  ignored when configured in slave mode, in master mode it contains the
+ *  address of the target peripheral. This is a 7-bit value unless 10-bit
+ *  addressing is configured and supported by the target.
+ *
+ *  When in master mode the operation consists of:
+ *   - Address the slave as a Master transmitter.
+ *   - Transmit data to the addressed slave.
+ *   - Generate a STOP condition if the specified `stop` field is true.
  *
  *  @param obj     The I2C object
  *  @param address 7-bit address (last bit is 0)
  *  @param data    The buffer for sending
  *  @param length  Number of bytes to write
- *  @param stop    Stop to be generated after the transfer is done
+ *  @param stop    If true, stop will be generated after the transfer is done
```

Comment:
Shouldn't this be removed from the parameter list as per the design doc ?

Reply:
`stop` stays